### PR TITLE
No json gang

### PR DESCRIPTION
**Pull request by liach**
Use data generators to make recipes! Be smart!

Note: The json recipes still exist in the mod jar, but we will create json with code instead of write json and suffer from all those error

Signed-off-by: liach <liach@users.noreply.github.com>

### CODE REVIEW

**Comment by liach** [comment]
Note that after merging this pr all future code needs to go to `src/code/java` instead of `src/main/java` because data generator cannot use code from main source set

File: build.gradle [hunk 1]
```diff
@@ -7,18 +7,18 @@ plugins {
...
+//		url = "https://minecraft.curseforge.com/api/maven"
+//	}
+//	maven {
+//		name = "GitHub"
+//		url = "https://maven.pkg.github.com"
+//	}
+//	maven {
+//		name = "JitPack"
+//		url = "https://jitpack.io"
+//	}
```
Comment (liach) [inline build.gradle]: These mavens make my configuartion slow as fuck (costs me extra 30 seconds configuring gradle) and breaks the source jar remapping, so I am removing them

**Comment by liach** [comment]
K updated this so that won't break other prs

**Comment by liach** [comment]
Conflict fixed, please check again @AK9Official

**Comment by Shnupbups** [comment]
no

**Comment by liach** [comment]
Oh a side note, this is just holding off till build.gradle is fixed.

File: src/main/java/com/fabriccommunity/spookytime/SpookyTime.java [hunk 2]
```diff
@@ -4,10 +4,15 @@
...
 import com.fabriccommunity.spookytime.common.SpookyEntities;
 import net.fabricmc.api.ModInitializer;
+import net.minecraft.util.Identifier;
 
-public class SpookyTime implements ModInitializer
+public final class SpookyTime implements ModInitializer
 {
 	public static final String MODID = "spookytime";
+	
+	public static Identifier id(String name) {
```
Comment (Shnupbups) [inline src/main/java/com/fabriccommunity/spookytime/SpookyTime.java:13]: This is good. I like this.

File: src/main/java/com/fabriccommunity/spookytime/SpookyTime.java [hunk 3]
```diff
@@ -4,10 +4,15 @@
 import com.fabriccommunity.spookytime.common.SpookyItems;
 import com.fabriccommunity.spookytime.common.SpookyEntities;
 import net.fabricmc.api.ModInitializer;
+import net.minecraft.util.Identifier;
 
-public class SpookyTime implements ModInitializer
+public final class SpookyTime implements ModInitializer
```
Comment (Shnupbups) [inline src/main/java/com/fabriccommunity/spookytime/SpookyTime.java:9]: Not a fan of final classes, but I suppose this is fine. Not like people are gonna extend a mod initializer.

File: src/data/java/com/fabriccommunity/spookytime/SpookyRecipeProvider.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	private CriterionConditions hasItem(ItemConvertible item)
+	{
+		return hasItem(item, 1);
+	}
+
+	private CriterionConditions hasItem(ItemConvertible item, int count)
+	{
+		return InventoryChangedCriterion.Conditions.items(ItemPredicate.Builder.create().item(item).count(NumberRange.IntRange.atLeast(count)).build());
+	}
+}
```
Comment (Shnupbups) [inline src/data/java/com/fabriccommunity/spookytime/SpookyRecipeProvider.java:139]: This whole class, and by extension the whole PR, looks... unwieldy and unnecessary. I don't see why we can't just use JSONs.

File: build.gradle [hunk 5]
```diff
@@ -49,6 +37,38 @@ dependencies {
...
+		compileClasspath += sourceSets.main.output
+		runtimeClasspath += sourceSets.main.compileClasspath
+		runtimeClasspath += sourceSets.main.output
+	}
+}
+
+def dataDir = project.file("$buildDir/datadump/")
+def dataCacheDir = new File(dataDir, "/.cache/")
+
+task generateData(type: JavaExec) {
```
Comment (Shnupbups) [inline build.gradle:52]: Will generating data not increase build time *significantly* if we have lots of data to generate?

File: src/data/java/com/fabriccommunity/spookytime/SpookyData.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+public final class SpookyData
+{
+	public static void main(String... args) throws IOException
+	{
+		Path path = Paths.get(args[0]);
+		DataGenerator dataGenerator = new DataGenerator(path, Collections.emptySet());
+		
+		dataGenerator.install(new SpookyRecipeProvider(dataGenerator));
+		
+		dataGenerator.run();
```
Comment (Shnupbups) [inline src/data/java/com/fabriccommunity/spookytime/SpookyData.java:20]: I don't get how this is better than just writing the JSONs directly? It's not like we'll have thousands of them that need to be created en masse.

**Review by Shnupbups — COMMENTED** [review]
> 
> 
> Note that after merging this pr all future code needs to go to `src/code/java` instead of `src/main/java` because data generator cannot use code from main source set

Also not a fan of this part ^

File: build.gradle (hunk 5 above)
Comment (liach) [reply build.gradle:52]: This should be cacheable. I will test more on the input and output of the task to make sure gradle can properly cache it.

File: src/data/java/com/fabriccommunity/spookytime/SpookyData.java (hunk 6 above)
Comment (liach) [reply src/data/java/com/fabriccommunity/spookytime/SpookyData.java:20]: Will add this back when we have a lot of jsons

**Pull request closed without merging**